### DISCUSSION TITLE
Different training size for benchmarking

### DISCUSSION THREAD

**Discussion by knc6** [janosh/matbench-discovery#69] [Q&A] [answered]
Hi,

I heard about this project at MRS Fall last week and I am looking at the matbench discovery page. Shouldn't training size (e.g. 1.6 million vs 133 k)  be the same for a fair comparison, else the model might be considered overfitting? This is machine learning 101.
![image](https://github.com/janosh/matbench-discovery/assets/16902896/1a83ca12-d6c2-44f8-baec-e9e4dc9018fd)

**Comment by janosh** [comment, marked as answer]
We ultimately don't care about the typical strict architecture comparison found in other ML benchmarks. We care about measuring how good ML (_any form_ of ML) is at OOD materials stability prediction. If some models (interatomic potentials) are trained on forces and therefore can leverage more of the maximum training set released with our benchmark (the entirety of the MP [v2022.10.28](https://docs.materialsproject.org/changes/database-versions#v2022.10.28) database version) then that's a genuine advantage of force-full models for the real-world application we care about and we want our benchmark to reflect that.

In short, we want to provide a walled garden for asking system-level questions which a traditional ML benchmark is too rigid to answer. I believe we succeeded at that. Matbench Discovery clearly demonstrated that universal interatomic potentials emulating DFT relaxation are the winning methodology for high-throughput OOD materials stability prediction.

**Comment by janosh** [comment]
Also, training set size and overfitting are two different concepts which you seem to be conflating. What's more, the empirical evidence suggests overfitting is a non-issue with large models.

**Comment by JonathanSchmidt1** [comment]
@janosh As gnome is there with an unknown large dataset, do you think it would make sense to add one of our CGAT models based on Alexandria (we would have to remove the wbm data first). However we usually only train our models for EHull prediction as that is the only target required for materials discovery. Of course one issue with that is that our convex hull is much more complete which will increase some errors.

**Reply by janosh** [reply]
absolutely, i'd love to have CGAT in the leaderboard! in fact, we're currently training a MACE and CHGNet from scratch and in fine-tuning mode on Alexandria at Radical. of course, the first thing we'll be trying those models on is matbench discovery 😄 
will be a good opportunity to gather empirical evidence on the effects of data leakage on this benchmark as we'll train once on all of Alex and once on the subset excluding all structural prototypes with matching compositions between WBM and Alex

**Reply by ml-evs** [reply]
Just following this up now that Orb has gone to the top of the leaderboard with Alexandria data in its training set (which fully contains WBM(?)), is there a recommended way to report whether you have/haven't excluded the WBM prototypes?

**Reply by ml-evs** [reply]
(Also wondering out loud to @JonathanSchmidt1, we could stick an OPTIMADE API on the WBM set on Materials Cloud (no need to host it yourselves) if that would make it easier to search/verify)

**Reply by CompRhys** [reply]
By my count the final prototype overlap with alex is about 10% of what we call the unique prototype set of WBM. I have an internal list of things to exclude from test set when training on Alex. We can look at making that public and having a leaderboard for it. One point to raise however is that when training on Alex the resolving power of the benchmark to look at architectural choices is reduced, I would prefer that we continue to get MP (MPtrj) submissions for MBD here or people train two models as in ORB so that we have better signal to noise. WBM is not OOD to Alex in the same way as it is to MP.

**Reply by ml-evs** [reply]
Makes sense, I think a slider/box similar to "is proprietary" for "guaranteed no data leakage" would be nice, but don't have a good intuition for whether this is actually a big effect (especially if you think the overlap is only 10%, would've guessed much higher -- though perhaps that 10% is biased towards the lowest energy structures anyway which could skew things?)

You can make the same argument about the proprietary datasets ofc, which we have no clue about

**Reply by CompRhys** [reply]
I feel the extra data and no extra data are the right buckets to toggle. Too many toggles could ruin the signal to noise. Also not sure about a toggle that adjusts the test set, if we went down that route we probably would just change the display subset to exclude the overlap. Every benchmark will be imperfect to some extent.

Honestly the results barely shifted when I looked at them for mattersim vs gnome as we know mattersim uses alex so may have this leakage.Maybe could impact mattersim vs ORB but to change that quickly after they release a new SOTA according to the metric we privileged in MBD might be unfair.

**Reply by JonathanSchmidt1** [reply]
Most if not the whole WBM dataset should be in Alexandria so training on it without removing it is not only not ouf of distribution but it is basically testing on your training set.
I am not sure if people are aware of this fact, which is also partially our fault as until this recent preprint https://papers.ssrn.com/sol3/papers.cfm?abstract_id=4893023 we never had a paper mentioning Alexandria ^^.
What I could suggest is that for training on Alexandria we could take one of the more recent batches of calculations ~ (we also have some new ones coming up) and say train on the rest and test on "this part". But this would be a new benchmark. E.g. the last batch that was published would be ~500k calculations and 27k new stable materials.

@ml-evs maybe ask Miguel about the optimade api if you would like to do that. I was not involved in that work so I cannot really comment.

I have been extremely busy lately but I hope I can also add some model to the benchmark in the next month. I could also try to separate the wbm dataset from alexandria and we can link to it somewhere for people who would like to use it for training.

**Reply by JonathanSchmidt1** [reply]
@CompRhys Do you mean that alexandria prototypes have a 10% overlap with WBM? that does not sound right to me, that would mean we forgot to upload sth maybe.

**Reply by CompRhys** [reply]
> @CompRhys Do you mean that alexandria prototypes have a 10% overlap with WBM? that does not sound right to me, that would mean we forgot to upload sth maybe.

Other way around, of the wbm structures if I look at the prototype+element labels for wyckoff positions then for around 10% of WBM I can find a similar structure in Alex.

**Reply by ml-evs** [reply]
> What I could suggest is that for training on Alexandria we could take one of the more recent batches of calculations ~ (we also have some new ones coming up) and say train on the rest and test on "this part". But this would be a new benchmark. E.g. the last batch that was published would be ~500k calculations and 27k new stable materials.

This would be nice, like the time splits suggested in https://joss.theoj.org/papers/10.21105/joss.05618 (could also be nicely handled with OPTIMADE's `last_modified` field)

**Reply by CompRhys** [reply]
[wbm_leaked_to_alex.csv](https://github.com/user-attachments/files/16870987/wbm_leaked_to_alex.csv)
These are the WBM ids I think have protostructure matches in Alex using the definition of protostructure from here: https://arxiv.org/abs/2309.16454. In my opinion removing these from the test set allows all of Alex as it exists now (3.2M structures) to be trained on with reduced leakage. It's an impossible task to avoid leakage entirely but the protostructure in my opinion provides a sufficiently discriminative bucket (although using fairly loose symprec) for grouping data that is much faster than the alternative of pairwise structure matching. 

The corresponding Alex ids are: 
[alex_leakage_ids_to_wbm.csv](https://github.com/user-attachments/files/16871039/alex_leakage_ids_to_wbm.csv)

The two lists are not the same length due to duplicated relaxed protostructures in WBM (for the unique protostructure subset shown as the main leaderboard only the lowest energy per protostructure is kept).

**Reply by JonathanSchmidt1** [reply]
@CompRhys  Thank you, I will have a look next week. Did you compare with the geoopt trajectory set endpoint or the relaxed set?

**Reply by JonathanSchmidt1** [reply]
> This would be nice, like the time splits suggested in https://joss.theoj.org/papers/10.21105/joss.05618 (could also be nicely handled with OPTIMADE's `last_modified` field)

Is there by chance already an endpoint for the download of full datasets through OPTIMADE by now.?

**Reply by CompRhys** [reply]
> @CompRhys Thank you, I will have a look next week. Did you compare with the geoopt trajectory set endpoint or the relaxed set?

I believe the geoopt trajectory set end point, the structures for that and the relaxed set will surely be identical though? i'd have guess the relaxed set is just statics on the trajectory end? This list ignores the energy. 

If you want to repeat the analysis I did, the sketch is something like:

```python
from aviary.wren.utils import get_protostructure_label_from_spglib
from pymatviz.enums import Key

# get dfs of wbm final and alex final structures with ids

df_wbm[Key.protostructure] = df_wbm[Key.final_struct].map(get_protostructure_label_from_spglib)
df_alex[Key.protostructure] = df_alex[Key.final_struct].map(get_protostructure_label_from_spglib)

df_wbm_drop = df_wbm[df_wbm[Key.protostructure].isin(df_alex[Key.protostructure])]
wbm_leakage_ids = pd.Series(df_wbm_drop[Key.mat_id].unique(), name=Key.mat_id)
wbm_leakage_ids.to_csv("wbm_leakage_ids_to_alex.csv", index=False)

df_alex_leaked = df_alex[df_alex[Key.protostructure].isin(df_wbm_drop[Key.protostructure])]
alex_leakage_ids = pd.Series(df_alex_leaked[Key.mat_id].unique(), name=Key.mat_id)
alex_leakage_ids.to_csv("alex_leakage_ids_from_wbm.csv", index=False)
```

Honestly I was very glad that there was this little leakage from WBM to Alex as it meant the MatterSim result was likely to not just have trained on all the test set. MBD v1 as a benchmark does suffer from this shifting window effect that as more data is generated over time we would expect that the data will eventually be within distribution c.f. [it being observably OOD based on increasing batch errors from 1 to 5 when trained only on MP](https://matbench-discovery.materialsproject.org/preprint#fig:rolling-mae-vs-hull-dist-wbm-batches-models).
